core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -504,17 +500,34 @@ public void run() {
 
   @Override
   public final void cancel(final Status reason) {
-    Substream noopSubstream = new Substream(0 /* previousAttempts doesn't matter here */);
-    noopSubstream.stream = new NoopClientStream();
-    Runnable runnable = commit(noopSubstream);
+    if (isHedging) {
+      Substream noopSubstream = new Substream(0 /* previousAttempts doesn't matter here */);
+      noopSubstream.stream = new NoopClientStream();
+      Runnable runnable = commit(noopSubstream);
 
-    if (runnable != null) {
+      if (runnable != null) {
+        synchronized (lock) {
+          state = state.substreamDrained(noopSubstream);
+        }
+        runnable.run();
+        safeCloseMasterListener(reason, RpcProgress.PROCESSED, new Metadata());
+        return;
+      }
+    } else {
+      Optional<Substream> substreamToCommit;
       synchronized (lock) {
-        state = state.substreamDrained(noopSubstream);
+        substreamToCommit = state.drainedSubstreams.stream().findFirst();
+      }
+      if (substreamToCommit.isPresent()) {
+        Substream substream = substreamToCommit.get();
+        Runnable runnable = commit(substream);
+        if (runnable != null) {
+          runnable.run();
+//          substream.stream.cancel(reason);
```

Comment:
I'm at a bit of a loss as to how to proceed here.

This is an attempt at committing an in-progress stream (please correct if it's wrong) rather than the noop stream for non-hedging retries. Committing an in-progress stream seems to prevent the committed stream from being cancelled, which breaks the contract with ManagedChannelImpl. Adding the cancel after the commit (i.e. uncommenting this line) fixes the contract but prevents the flush and causes the buffer leak.

Am I missing something about the usage here, or is this not possible without a larger rework for retriable streams? Thanks for the patience in any case.

Reply:
> Adding the cancel after the commit (i.e. uncommenting this line) fixes the contract but prevents the flush and causes the buffer leak.

Only because you added `framer().dispose();` to cancel. If we keep the "cancel doesn't impact the framer," then this would still work.

I know it seems really strange that we can _avoid_ a memory leak by _not_ closing a resource. But we're essentially modeling this as "that's the other thread's problem," which has some wisdom in it.

Reply:
Thanks, I tried removing the `framer().dispose()` from cancel but the leak still occurs if we keep the `stream.cancel()` but we break the ManagedChannelImpl + DelayedClientTransport contract if we omit it, i.e. `./gradlew :grpc-core:test --tests io.grpc.internal.ManagedChannelImplIdlenessTest.newCallRefreshesIdlenessTimer`. To be clear: we should be cancelling the committed stream here, right?

Do I need to move the `framer().dispose()` call to another location to flush and free the buffer if we keep the cancel?